In [ ]:
import pandas as pd
from google.colab import files

uploaded = files.upload()

df = pd.read_csv("gss.csv")

print("Dataset shape:", df.shape)
df.head()

Saving gss.csv to gss.csv
Dataset shape: (75699, 61)


,ETHNIC,BALLOT,WRKSTAT,OCC10,MARITAL,SIBS,AGE,EDUC,SEX,RACE,...,NEWSKILL,LOWPAY,TEMPWORK,MORTRAVL,ECONSUP1,WORECSUP,SATJOBHV,ID_,YEAR,ONET_SOC_CODE
0,NaN,NaN,1.0,520.0,5.0,3.0,23.0,16.0,2.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,1972.0,13-1022.00
1,14.0,NaN,5.0,7700.0,1.0,4.0,70.0,10.0,1.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2.0,1972.0,51-1011.00
2,15.0,NaN,2.0,4920.0,1.0,5.0,48.0,12.0,2.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.0,1972.0,41-9022.00
3,37.0,NaN,1.0,800.0,1.0,5.0,27.0,17.0,2.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.0,1972.0,13-2011.00
4,27.0,NaN,7.0,5020.0,1.0,2.0,61.0,12.0,2.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5.0,1972.0,43-2021.00


In [ ]:
print("Missing ONET_SOC_CODE:")
print(df["ONET_SOC_CODE"].isna().sum())

print("\nPercentage missing:")
print(df["ONET_SOC_CODE"].isna().mean() * 100)

Missing ONET_SOC_CODE:
5616

Percentage missing:
7.418856259659969


In [ ]:
onet_missing_by_year = (
    df.groupby("YEAR")["ONET_SOC_CODE"]
    .apply(lambda x: x.isna().sum())
)

print(onet_missing_by_year)

YEAR
1972.0    350
1973.0    252
1974.0    235
1975.0    252
1976.0    201
1977.0    281
1978.0    139
1980.0    172
1982.0    175
1983.0    139
1984.0     99
1985.0    122
1986.0    116
1987.0    183
1988.0     92
1989.0    116
1990.0     78
1991.0    101
1993.0     87
1994.0    153
1996.0    123
1998.0    169
2000.0    147
2002.0    116
2004.0    157
2006.0    235
2008.0     76
2010.0    126
2012.0    103
2014.0    133
2016.0    129
2018.0    100
2021.0    204
2022.0    242
2024.0    213
Name: ONET_SOC_CODE, dtype: int64


In [ ]:
missing_onet = df[df["ONET_SOC_CODE"].isna()]

print("Rows with missing ONET_SOC_CODE:", len(missing_onet))

print("\nOCC10 values among rows missing ONET_SOC_CODE:")
print(missing_onet["OCC10"].value_counts(dropna=False).head(20))

Rows with missing ONET_SOC_CODE: 5616

OCC10 values among rows missing ONET_SOC_CODE:
OCC10
NaN       5337
9830.0     279
Name: count, dtype: int64


In [ ]:
# Look at all rows where OCC10 = 9830
occ9830 = df[df["OCC10"] == 9830]

print("Number of OCC10 = 9830 rows:", len(occ9830))

# Look at the other occupation-related variables
print(occ9830[["YEAR", "OCC10", "ONET_SOC_CODE"]].head(20))

Number of OCC10 = 9830 rows: 279
        YEAR   OCC10 ONET_SOC_CODE
32    1972.0  9830.0           NaN
60    1972.0  9830.0           NaN
148   1972.0  9830.0           NaN
179   1972.0  9830.0           NaN
442   1972.0  9830.0           NaN
795   1972.0  9830.0           NaN
827   1972.0  9830.0           NaN
829   1972.0  9830.0           NaN
832   1972.0  9830.0           NaN
833   1972.0  9830.0           NaN
895   1972.0  9830.0           NaN
972   1972.0  9830.0           NaN
1121  1972.0  9830.0           NaN
1149  1972.0  9830.0           NaN
1470  1972.0  9830.0           NaN
1545  1972.0  9830.0           NaN
1720  1973.0  9830.0           NaN
1957  1973.0  9830.0           NaN
2192  1973.0  9830.0           NaN
2250  1973.0  9830.0           NaN


In [ ]:
print(
    "OCC10 = 9830 with missing ONET:",
    occ9830["ONET_SOC_CODE"].isna().sum()
)

print(
    "OCC10 = 9830 with ONET:",
    occ9830["ONET_SOC_CODE"].notna().sum()
)

OCC10 = 9830 with missing ONET: 279
OCC10 = 9830 with ONET: 0


In [ ]:
print(
    df.loc[df["ONET_SOC_CODE"].isna(), "WRKSTAT"]
    .value_counts(dropna=False)
)

WRKSTAT
7.0    2586
6.0     938
1.0     870
5.0     569
8.0     290
4.0     136
2.0     136
3.0      49
NaN      42
Name: count, dtype: int64


In [ ]:
# Create a table showing the percentage of each work-status group
# that is missing an O*NET code

wrkstat_table = pd.crosstab(
    df["WRKSTAT"],
    df["ONET_SOC_CODE"].isna(),
    normalize="index"
) * 100

wrkstat_table.columns = ["Has_ONET_%", "Missing_ONET_%"]

print(wrkstat_table.round(2))

         Has_ONET_%  Missing_ONET_%
WRKSTAT                            
1.0           97.63            2.37
2.0           98.25            1.75
3.0           96.97            3.03
4.0           95.14            4.86
5.0           95.13            4.87
6.0           58.81           41.19
7.0           76.62           23.38
8.0           83.33           16.67


ONET_SOC_CODE is missing for 5,616 rows.
5,337 (95%) of those also have missing OCC10.
279 (5%) have OCC10 = 9830, meaning “Military, rank not specified.”
Missing O*NET codes are not completely random because the rate changes by work status.
Missingness is especially high for people in school (41.19%), keeping house (23.38%), and other (16.67%).
For the analysis, we should keep these values as missing rather than guessing an occupation and exclude them when a valid O*NET occupation is required.

Next is Task #2: Identify the variables of interest for job fit.

The main idea is: compare what people value in a job with how satisfied they are with their current job.

In [17]:
# Variables we think could be relevant to job satisfaction
candidate_vars = [
    "SATJOB",
    "JOBSAT",
    "JOBMEANS",
    "JOBKEEP",
    "JOBRISE",
    "JOBHONOR",
    "JOBOFF",
    "JOBINTER",
    "JOBINDEP",
    "JOBRESP",
    "JOBPEOP",
    "JOBHELP",
    "JOBSOC",
    "JOBACCMP",
    "JOBSAFE",
    "IMPPROMO",
    "ILIKEJOB",
    "TIMEPDWK",
    "WRKEARN",
    "WRKENJOY",
    "WRKSTAT",
    "OCC10",
    "AGE",
    "EDUC",
    "SEX",
    "RACE",
    "INCOME",
    "MARITAL",
    "YEAR"
]

# Keep only variables that actually exist in the dataset
candidate_vars = [v for v in candidate_vars if v in df.columns]

coverage = pd.DataFrame({
    "Valid": df[candidate_vars].notna().sum(),
    "Missing": df[candidate_vars].isna().sum(),
    "Percent Valid": df[candidate_vars].notna().mean() * 100
})

coverage = coverage.sort_values("Percent Valid", ascending=False)

print(coverage)

          Valid  Missing  Percent Valid
YEAR      75699        0     100.000000
WRKSTAT   75652       47      99.937912
MARITAL   75633       66      99.912813
SEX       75568      131      99.826946
RACE      75527      172      99.772784
EDUC      75413      286      99.622188
AGE       74829      870      98.850711
OCC10     70362     5337      92.949709
INCOME    66370     9329      87.676191
SATJOB    54131    21568      71.508210
JOBMEANS  21814    53885      28.816761
WRKEARN    5567    70132       7.354126
WRKENJOY   5534    70165       7.310533
JOBSAT     3656    72043       4.829654
TIMEPDWK   3453    72246       4.561487
ILIKEJOB   1359    74340       1.795268
JOBKEEP     918    74781       1.212698
JOBHELP     918    74781       1.212698
JOBINTER    915    74784       1.208735
JOBOFF      915    74784       1.208735
JOBRISE     915    74784       1.208735
JOBHONOR    914    74785       1.207414
JOBSOC      914    74785       1.207414
JOBPEOP     914    74785       1.207414


Task #3: Missing values and outliers after selecting your variables.

In [19]:
model_vars = [
    "SATJOB",
    "AGE",
    "SEX",
    "RACE",
    "EDUC",
    "INCOME",
    "MARITAL",
    "WRKSTAT",
    "OCC10",
    "YEAR"
]
df["model_complete"] = df[model_vars].notna().all(axis=1)

print("Total rows:", len(df))
print("Usable rows:", df["model_complete"].sum())
print("Percent usable:", df["model_complete"].mean() * 100)

Total rows: 75699
Usable rows: 45837
Percent usable: 60.55165854238497


In [20]:
# Check minimum and maximum values for numerical variables
numeric_vars = ["AGE", "EDUC", "INCOME", "SATJOB", "YEAR"]

for var in numeric_vars:
    print("\n", var)
    print("Minimum:", df[var].min())
    print("Maximum:", df[var].max())


 AGE
Minimum: 18.0
Maximum: 89.0

 EDUC
Minimum: 0.0
Maximum: 20.0

 INCOME
Minimum: 1.0
Maximum: 12.0

 SATJOB
Minimum: 1.0
Maximum: 4.0

 YEAR
Minimum: 1972.0
Maximum: 2024.0


In [21]:
for var in numeric_vars:
    print("\n", var)
    print(df[var].value_counts().sort_index().head(20))


 AGE
AGE
18.0     297
19.0     938
20.0     978
21.0    1112
22.0    1197
23.0    1354
24.0    1340
25.0    1511
26.0    1492
27.0    1515
28.0    1592
29.0    1492
30.0    1621
31.0    1522
32.0    1625
33.0    1584
34.0    1614
35.0    1553
36.0    1560
37.0    1552
Name: count, dtype: int64

 EDUC
EDUC
0.0       189
1.0        50
2.0       161
3.0       273
4.0       329
5.0       414
6.0       889
7.0       906
8.0      2812
9.0      2209
10.0     3079
11.0     4046
12.0    22202
13.0     6188
14.0     8651
15.0     3484
16.0    10666
17.0     2571
18.0     3187
19.0     1166
Name: count, dtype: int64

 INCOME
INCOME
1.0       969
2.0      1460
3.0      1335
4.0      1207
5.0      1339
6.0      1263
7.0      1331
8.0      2319
9.0      7025
10.0     5404
11.0     5667
12.0    37051
Name: count, dtype: int64

 SATJOB
SATJOB
1.0    25711
2.0    20776
3.0     5372
4.0     2272
Name: count, dtype: int64

 YEAR
YEAR
1972.0    1613
1973.0    1504
1974.0    1484
1975.0    1490
1976.0    

In [22]:
print("AGE")
print(df["AGE"].value_counts().sort_index().head(20))
print(df["AGE"].value_counts().sort_index().tail(20))

AGE
AGE
18.0     297
19.0     938
20.0     978
21.0    1112
22.0    1197
23.0    1354
24.0    1340
25.0    1511
26.0    1492
27.0    1515
28.0    1592
29.0    1492
30.0    1621
31.0    1522
32.0    1625
33.0    1584
34.0    1614
35.0    1553
36.0    1560
37.0    1552
Name: count, dtype: int64
AGE
70.0    906
71.0    773
72.0    759
73.0    681
74.0    760
75.0    631
76.0    605
77.0    584
78.0    510
79.0    470
80.0    417
81.0    386
82.0    330
83.0    306
84.0    278
85.0    236
86.0    222
87.0    161
88.0    133
89.0    437
Name: count, dtype: int64


After selecting the variables for analysis, I checked the dataset for missing values and potential outliers. Of the 75,699 total observations, 45,837 observations were complete for all selected variables, giving 60.55% usable data. I also examined the ranges and frequency distributions of the numerical and ordinal variables. No obvious invalid values or extreme outliers were identified. For example, age values ranged from 18 to 89, while SATJOB contained values from 1 to 4. INCOME contained categories from 1 to 12 rather than exact dollar amounts. Because these values represent valid survey responses or categories, I did not remove them as outliers.


Task 4: EDA

In [23]:
df.groupby('SATJOB')['AGE'].mean()

,AGE
SATJOB,
1.0,44.531909
2.0,40.697459
3.0,38.813065
4.0,38.518024


In [24]:
df.groupby('SATJOB')['EDUC'].mean()

,EDUC
SATJOB,
1.0,13.413025
2.0,13.229592
3.0,13.007104
4.0,12.762262


In [25]:
df.groupby('SATJOB')['INCOME'].mean()

,INCOME
SATJOB,
1.0,10.613627
2.0,10.437410
3.0,9.937617
4.0,9.363458


In [26]:
pd.crosstab(df['SEX'], df['SATJOB'], normalize='index') * 100

SATJOB,1.0,2.0,3.0,4.0
SEX,,,,
1.0,46.897017,39.184913,9.863730,4.054339
2.0,47.982329,37.758803,9.959778,4.299090


In [27]:
pd.crosstab(df['RACE'], df['SATJOB'], normalize='index') * 100

SATJOB,1.0,2.0,3.0,4.0
RACE,,,,
1.0,49.334114,37.607958,9.200702,3.857226
2.0,39.630977,40.800416,13.422557,6.146050
3.0,43.016448,42.124338,10.816839,4.042375


In [28]:
pd.crosstab(df['MARITAL'], df['SATJOB'], normalize='index') * 100

SATJOB,1.0,2.0,3.0,4.0
MARITAL,,,,
1.0,51.514739,37.050763,8.381218,3.053279
2.0,55.985663,33.189964,7.275986,3.548387
3.0,46.996955,37.489621,10.462220,5.051204
4.0,43.815514,37.368973,11.949686,6.865828
5.0,37.245218,43.242396,13.428975,6.083412


In [29]:
pd.crosstab(df['WRKSTAT'], df['SATJOB'], normalize='index') * 100

SATJOB,1.0,2.0,3.0,4.0
WRKSTAT,,,,
1.0,48.928540,38.776641,9.111053,3.183766
2.0,45.324441,39.488676,10.559956,4.626928
3.0,47.569444,37.708333,10.000000,4.722222
4.0,33.088553,38.790497,16.501080,11.619870
5.0,66.666667,33.333333,0.000000,0.000000
6.0,33.333333,50.000000,16.666667,0.000000
7.0,47.465854,35.963427,10.825150,5.745569
8.0,33.333333,33.333333,0.000000,33.333333


In [35]:
df[['AGE', 'EDUC', 'INCOME', 'YEAR', 'SATJOB']].corr()['SATJOB']

,SATJOB
AGE,-0.149141
EDUC,-0.054710
INCOME,-0.106792
YEAR,0.003129
SATJOB,1.000000


In [30]:
df['WRKSTAT'].value_counts().sort_index()

,count
WRKSTAT,
1.0,36727
2.0,7750
3.0,1618
4.0,2796
5.0,11683
6.0,2277
7.0,11061
8.0,1740


In [31]:
df['OCC10'].value_counts().sort_index()

,count
OCC10,
10.0,395
20.0,361
30.0,12
40.0,36
50.0,402
...,...
9750.0,51
9800.0,70
9810.0,122


In [32]:
df['OCC10'].value_counts().head(10)

,count
OCC10,
5700.0,2788
4760.0,1579
2310.0,1555
3600.0,1464
9130.0,1389
4700.0,1327
4230.0,1253
3255.0,1175
4110.0,1118


In [33]:
top_occ = df['OCC10'].value_counts().head(10).index

pd.crosstab(
    df[df['OCC10'].isin(top_occ)]['OCC10'],
    df[df['OCC10'].isin(top_occ)]['SATJOB'],
    normalize='index'
) * 100

SATJOB,1.0,2.0,3.0,4.0
OCC10,,,,
2310.0,59.136546,31.827309,6.827309,2.208835
3255.0,53.389831,34.745763,8.716707,3.147700
3600.0,47.358491,38.018868,10.471698,4.150943
4110.0,36.938032,40.218712,14.459295,8.383961
4220.0,37.515078,42.702051,13.268999,6.513872
4230.0,40.519187,39.277652,13.544018,6.659142
4700.0,50.896414,35.458167,10.059761,3.585657
4760.0,40.176211,42.555066,11.189427,6.079295
5700.0,51.314459,36.248736,8.796764,3.640040


In [34]:
df.groupby('YEAR')['SATJOB'].mean()

,SATJOB
YEAR,
1972.0,1.697034
1973.0,1.672217
1974.0,1.713818
1975.0,1.630901
1976.0,1.659916
1977.0,1.682250
1978.0,1.666406
1980.0,1.753612
1982.0,1.760054


To identify which features appear most related to job satisfaction, I compared each selected variable with `SATJOB`. Among the numeric variables, `AGE` had the strongest association with job satisfaction (correlation = -0.149), followed by `INCOME` (correlation = -0.107). `EDUC` had a weaker association (correlation = -0.055), while `YEAR` showed essentially no association (correlation = 0.003). For categorical variables, `MARITAL`, `WRKSTAT`, and `OCC10` showed noticeable differences in job satisfaction across groups. `RACE` showed some differences, while `SEX` had very similar job-satisfaction distributions across its categories. Based on this exploratory analysis, `AGE`, `INCOME`, `MARITAL`, `WRKSTAT`, and `OCC10` appear to be the most promising features for further analysis of job satisfaction. These relationships are descriptive associations and do not imply that these variables cause differences in job satisfaction.

My analysis focused on job satisfaction (`SATJOB`) and examined demographic, economic, employment, occupation, and time-related variables. The variables selected for analysis were `AGE`, `SEX`, `RACE`, `EDUC`, `INCOME`, `MARITAL`, `WRKSTAT`, `OCC10`, and `YEAR`.

Overall, the dataset is suitable for analyzing job satisfaction because it contains a large number of observations and includes relevant demographic and employment information. However, there are some limitations. Of the 75,699 total observations, 45,837 were complete for all selected variables, meaning that approximately 60.55% of the observations could be used without missing values. In addition, `INCOME` is recorded as categories rather than exact income amounts, and `OCC10` contains a large number of occupation codes. Despite these limitations, the dataset contains enough relevant information to explore factors associated with job satisfaction.
